File: intern/1261.find-elements-in-a-contaminated-binary-tree.js
/*
 * @lc app=leetcode id=1261 lang=javascript
 *
 * [1261] Find Elements in a Contaminated Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
class FindElements {
    #root;

    static #LEFT = 0;
    static #RIGHT = 1;

    constructor(root) {
        this.#root = root;
    }

    /**
     * @param {number} target
     * @return {boolean}
     */
    find(target) {
        let current = target;
        const path = [];

        while (current > 0) {
            if (current % 2 === 0) {
                current = (current - 2) / 2;
                path.push(FindElements.#RIGHT);
            } else {
                current = (current - 1) / 2;
                path.push(FindElements.#LEFT);
            }
        }

        let currentNode = this.#root;
        for (let i = path.length - 1; i >= 0; i--) {
            const direction = path[i];
            if (direction === FindElements.#LEFT) {
                currentNode = currentNode.left;
            } else if (direction === FindElements.#RIGHT) {
                currentNode = currentNode.right;
            }

            if (currentNode === null) {
                return false;
            }
        }

        return true;
    }
}

/**
 * Your FindElements object will be instantiated and called as such:
 * var obj = new FindElements(root)
 * var param_1 = obj.find(target)
 */
// @lc code=end

File: src/2025/1261.find-elements-in-a-contaminated-binary-tree.cpp
/*
 * @lc app=leetcode id=1261 lang=cpp
 *
 * [1261] Find Elements in a Contaminated Binary Tree
 */

#include <unordered_set>

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// @lc code=start
class FindElements {
 public:
  FindElements(TreeNode* root) : root_(root) {
    root_->val = 0;
    elements.insert(0);
    recoverRoot(root_);
  }

  bool find(int target) { return elements.contains(target); }

 private:
  void recoverRoot(TreeNode* root) {
    elements.insert(root->val);
    if (root->left) {
      root->left->val = root->val * 2 + 1;
      recoverRoot(root->left);
    }
    if (root->right) {
      root->right->val = root->val * 2 + 2;
      recoverRoot(root->right);
    }
  }

 private:
  TreeNode* root_;
  std::unordered_set<int> elements;
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */
// @lc code=end
